Isolated Content Serialization

stevewgh opened the pull request:
My first effort at adding support for custom serializers #566 . I'm interested in hearing what you think, getting feedback etc. Thanks.

### Overview of Changes 
- Isolated all content serialization to the IContentSerializer interface
- Added JsonContentSerializer and XmlContentSerializer implmentations
- RefitSettings allows the IContentSerializer to be specified but still defaults to JsonContentSerializer for back-compat reasons
- .NET Standard 1.4 has a new package dependency on System.Xml.XmlSerializer

### Tests
- Tests modified to exercise Json and Xml serialization
- XmlContentSerializer unit test added
- Some existing tests have been modified to use both JsonContentSerializer and XmlContentSerializer using the an xunit theory

### How to use
```c#
// 1. uses Json serialization - no change
var service = RestService.For<IService>("https://localhost");

// 2. uses Json serialization and passes the JsonSerializerSettings to the JsonContentSerializer
var service = RestService.For<IService>("https://localhost",
    new RefitSettings
    {
        JsonSerializerSettings = new JsonSerializerSettings {Formatting = Formatting.Indented}
    });

// 3. equivalent of 2. 
var service = RestService.For<IService>("https://localhost",
    new RefitSettings
    {
        ContentSerializer = new JsonContentSerializer(new JsonSerializerSettings {Formatting = Formatting.Indented})
    });

// 4. uses Xml serialization
var service = RestService.For<IService>("https://localhost",
    new RefitSettings
    {
        ContentSerializer = new XmlContentSerializer()
    });

// 5. uses Xml serialization with additional configuration
var service = RestService.For<IService>("https://localhost",
    new RefitSettings
    {
        ContentSerializer = new XmlContentSerializer(
            new XmlContentSerializerSettings
            {
                XmlReaderWriterSettings = new XmlReaderWriterSettings(
                    new XmlReaderSettings
                    {
                        ConformanceLevel = ConformanceLevel.Document
                    })
            })
    });
```

### Possible further changes
- Should the ```RefitSettings.JsonSerializerSettings``` property as marked as ```[Obsolete]```? It should only be used with the JsonContentSerializer.
- Simplify how the ContentSerializer property is set. One idea is to add a ```RefitSettings``` extension class - something like this:

```c#
public static class RefitSettingsJsonContentExtensions
{
    public static RefitSettings UseJsonSerialization(this RefitSettings settings, JsonSerializerSettings jsonSerializerSettings)
    {
        settings.ContentSerializer = new JsonContentSerializer(jsonSerializerSettings);
        return settings;
    }
}
```
The extension class would be used like this:
```c#
var service = RestService.For<IService>("https://localhost", new RefitSettings().UseJsonSerialization(new JsonSerializerSettings {Formatting = Formatting.Indented}));
```

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/reactiveui/refit?pullRequest=571) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/reactiveui/refit?pullRequest=571) before we can accept your contribution.<br/><br/>:x: stevewgh [sign now](https://cla.dotnetfoundation.org/reactiveui/refit?pullRequest=571)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/reactiveui/refit?pullRequest=571) it.</sub>

onovotny commented:
Thanks for this!

> Should the RefitSettings.JsonSerializerSettings property as marked as [Obsolete]? It should only be used with the JsonContentSerializer.

Yes, I would say so, but with a warning, not an error, and the message should point to the new type and docs.

I'm not keen on using an extension method to configure a new instance of the settings object -- @paulcbetts any thoughts?

stevewgh commented:
Glad to help! 

I've rebased my branch on master and made some further changes. 

- Marked RefitSettings.JsonSerializerSettings as obsolete
- Marked the BodySerializationMethod.Json enum as obsolete, suggest using the newly added BodySerializationMethod.Serialized instead
- Updated the README document with examples of how to use Refit with XML

I hope this is OK? 

Thanks.

onovotny commented on Refit/RefitSettings.cs:
```diff
@@ -13,21 +12,44 @@ namespace Refit
+    {
+        HttpContent Serialize(object item);
+
+        Task<object> DeserializeAsync(HttpContent content, Type objectType);
```
Can this be `Task<T> DeserializeAsync<T>(HttpContent content)` to improve type safety?

onovotny replied:
Also, why do we need both a sync & async deserialize? if we need it, do we also need an async serialize?

onovotny commented on Refit.Tests/MultipartTests.cs:
```diff
@@ -429,7 +429,7 @@ public async Task MultipartUploadShouldWorkWithAnObject(Type contentSerializerTy
                     Assert.Null(parts[0].Headers.ContentDisposition.FileName);
                     Assert.Equal(mediaType, parts[0].Headers.ContentType.MediaType);
-                    var result0 = serializer.Deserialize<ModelObject>(await parts[0].ReadAsStringAsync());
+                    var result0 = await serializer.DeserializeAsync<ModelObject>(parts[0]);
```
needs `ConfigureAwait(false)`

onovotny commented on Refit/ValidationApiException.cs:
```diff
@@ -27,7 +27,6 @@ public static ValidationApiException Create(ApiException exception)
         /// </summary>
-        public new ProblemDetails Content => GetContentAs<ProblemDetails>();
-
+        public new ProblemDetails Content => GetContentAsAsync<ProblemDetails>().ConfigureAwait(false).GetAwaiter().GetResult();
```
This can deadlock. Async calls need to be awaited, thus cannot really be a property.

onovotny commented on Refit/RequestBuilderImplementation.cs:
```diff
@@ -574,12 +575,12 @@ void AddMultipartItem(MultipartFormDataContent multiPartContent, string fileName
                         {
-                            AddMultipartItem(multiPartContent, itemName, parameterName, item);
+                            await AddMultipartItemAsync(multiPartContent, itemName, parameterName, item);
                         }
```
needs `ConfigureAwait(false)`

onovotny commented on Refit/RequestBuilderImplementation.cs:
```diff
@@ -574,12 +575,12 @@ void AddMultipartItem(MultipartFormDataContent multiPartContent, string fileName
                     else
                     {
-                        AddMultipartItem(multiPartContent, itemName, parameterName, itemValue);
+                        await AddMultipartItemAsync(multiPartContent, itemName, parameterName, itemValue);
```
needs `ConfigureAwait(false)`

onovotny commented on Refit/RequestBuilderImplementation.cs:
```diff
@@ -574,12 +575,12 @@ void AddMultipartItem(MultipartFormDataContent multiPartContent, string fileName
                         foreach (var item in enumerable)
                         {
-                            AddMultipartItem(multiPartContent, itemName, parameterName, item);
+                            await AddMultipartItemAsync(multiPartContent, itemName, parameterName, item);
```
needs `ConfigureAwait(false)`

onovotny commented on Refit/RequestBuilderImplementation.cs:
```diff
@@ -670,7 +671,7 @@ void AddMultipartItem(MultipartFormDataContent multiPartContent, string fileName
 
                 var factory = BuildRequestFactoryForMethod(restMethod, client.BaseAddress.AbsolutePath, restMethod.CancellationToken != null);
-                var rq = factory(paramList);
+                var rq = await factory(paramList);
```
needs `ConfigureAwait(false)`

onovotny commented on Refit/RequestBuilderImplementation.cs:
```diff
@@ -212,23 +214,23 @@ void AddMultipartItem(MultipartFormDataContent multiPartContent, string fileName
 
                 var factory = BuildRequestFactoryForMethod(restMethod, client.BaseAddress.AbsolutePath, restMethod.CancellationToken != null);
-                var rq = factory(paramList);
+                var rq = await factory(paramList);
```
needs `ConfigureAwait(false)`

stevewgh commented:
Good spot with the the missing ConfigureAwait's, I'll add them in.

The ValidationApiException.Content property is a bit trickier. If I change this factory method to `async Task<ValidationApiException>`
https://github.com/reactiveui/refit/blob/c14b2c265e3ebb1f72915ec44e8ea73ef136fed4/Refit/ValidationApiException.cs#L22-L25

I could then set the Content property from within the `Create` method and the property would avoid deadlocks.

We already have an async context where the `ValidationApiException.Create` is called, so I think this could work.

onovotny commented on Refit/ValidationApiException.cs:
```diff
@@ -14,19 +15,24 @@ public class ValidationApiException : ApiException
-            return new ValidationApiException(exception);
+            return new ValidationApiException(exception)
+            {
+                Content = await exception.GetContentAsAsync<ProblemDetails>()
```
Needs `ConfigureAwait(false)`

onovotny commented on Refit/RefitSettings.cs:
```diff
@@ -13,21 +12,42 @@ namespace Refit
 
+    public interface IContentSerializer
+    {
+        Task<HttpContent> SerializeAsync(object item);
```
Can we make this generic, so it's `SerializeAsync<T>(T item)` that could prevent a box and allows for a custom serializer that's optimized for a particular type.

stevewgh replied:
That was an easier change than I was expecting! The method swapped over to a generic with no issues.

onovotny commented:
Looks good! One last thing and then I think it can be merged.

stevewgh commented:
Hi @onovotny are the changes that I made in https://github.com/reactiveui/refit/pull/571/commits/cc05c3a9feea80200309f40beed9cbe4b304883f  ok?

richbryant commented:
This still doesn't make it easy to use one's own ContentSerializer, since the IContentSerializer interface is buried in RefitSettings.  I'm using Blazor, Json.NET is not supported under WASM, I can easily do my own deserialization based on returning HttpResponseMessage but posting is proving more than a little evil.

If I could simply throw in something that implements IContentSerializer, it'd be easy.  But I can't.  Why not?

stevewgh commented:
Hi @richbryant if you want to implement your own IContentSerializer then you can. Here's one which sends everything as Base64 strings and uses a non-existent (I think!) media type of text/base64

```cs
public class Base64ContentSerializer : IContentSerializer
{
    BinaryFormatter formatter = new BinaryFormatter();

    public async Task<T> DeserializeAsync<T>(HttpContent content)
    {
        var bytes = Convert.FromBase64String(await content.ReadAsStringAsync());

        using (var ms = new MemoryStream(bytes))
        {
            return (T)formatter.Deserialize(ms);
        }
    }

    public Task<HttpContent> SerializeAsync<T>(T item)
    {
        using (var stream = new MemoryStream())
        {
            formatter.Serialize(stream, item);
            var content = (HttpContent)new StringContent(Convert.ToBase64String(stream.ToArray()), System.Text.Encoding.UTF8, "text/base64");
            return Task.FromResult(content);
        }
    }
}

// use it like this

var client = RestService.For<IApi>("http://localhost:5000",
    new RefitSettings
    {
        ContentSerializer = new Base64ContentSerializer()
    });
```

The `IContentSerializer` interface is available in versions 4.6.68 onwards which is only available on the 'MyGet' feed for now. 

@onovotny do you have an ETA for when 4.6.68 be published to NuGet?
